cluster-operator/src/main/java/io/strimzi/operator/cluster/PlatformFeaturesAvailability.java

```diff
@@ -4,23 +4,138 @@
+        }).compose(supported -> {
+            pfa.setRoutes(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "build.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setBuilds(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "apps.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setApps(supported);
+            return checkApiAvailability(vertx, httpClient, client.getMasterUrl().toString(), "image.openshift.io", "v1");
+        }).compose(supported -> {
+            pfa.setImages(supported);
+            composeFuture.complete();
+        }, composeFuture);
+
+        composeFuture.setHandler(res -> {
+            if (res.succeeded())  {
+                pfaFuture.complete(pfa);
+            } else {
+                pfaFuture.fail(res.cause());
+            }
+        });
+
+        return pfaFuture;
+    }
+
+    private static OkHttpClient getOkHttpClient(KubernetesClient client)   {
+        if (client.isAdaptable(OkHttpClient.class)) {
+            return client.adapt(OkHttpClient.class);
+        } else {
+            log.error("Cannot adapt KubernetesClient to OkHttpClient");
+            throw new RuntimeException("Cannot adapt KubernetesClient to OkHttpClient");
+        }
+    }
+
+    private static Future<VersionInfo> getVersionInfo(Vertx vertx, KubernetesClient client)   {
+        Future<VersionInfo> fut = Future.future();
+
+        vertx.executeBlocking(request -> {
+            try {
+                request.complete(client.getVersion());
+            } catch (Exception e) {
+                log.error("Detection of Kuberetes version failed.", e);
+                request.fail(e);
+            }
+        }, fut.completer());
+
+        return fut;
+    }
+
+    private static Future<Boolean> checkApiAvailability(Vertx vertx, OkHttpClient httpClient, String masterUrl, String api, String version)   {
+        Future<Boolean> fut = Future.future();
+
+        vertx.executeBlocking(request -> {
+            try {
+                Boolean isSupported;
+
+                Response resp = httpClient.newCall(new Request.Builder().get().url(masterUrl + "apis/" + api + "/" + version).build()).execute();
+                if (resp.code() >= 200 && resp.code() < 300) {
+                    log.debug("{} returned {}. This API is supported.", resp.request().url(), resp.code());
+                    isSupported = true;
+                } else {
+                    log.debug("{} returned {}. This API is not supported.", resp.request().url(), resp.code());
+                    isSupported = false;
+                }
+
+                resp.close();
+                request.complete(isSupported);
+            } catch (Exception e) {
+                log.error("Detection of {}/{} API failed.", api, version, e);
+                request.fail(e);
```

Comment:
Do we really want to fail here? Or should we just complete with false? It would mean buggy detection wasn't a fatal error for the operator, merely that certain features would be unavailable.

Reply:
Good point. I passed it with `false` and updated the error message accordingly.